includes/data-port/class-sensei-data-port-manager.php

```diff
@@ -0,0 +1,201 @@
+	const OPTION_NAME = 'sensei-tools-jobs';
+
+	/**
+	 * An array of all in progress data port jobs. It has the following format:
+	 * {
+	 *
+	 *     @type string $id        Unique id for this job.
+	 *     @type int    $user_id   The user which initiatied this job.
+	 *     @type int    $time      When the job started.
+	 *     @type string $handler   The class which handles this job.
+	 * }
+	 *
+	 * @var array
+	 */
+	private $data_port_jobs;
+
+	/**
+	 * Tracks if the data port jobs have been updated.
+	 *
+	 * @var boolean
+	 */
+	private $has_changed;
+
+	/**
+	 * Instance of singleton.
+	 *
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * Fetches an instance of the class.
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Sensei_Data_Port_Manager constructor.
+	 */
+	private function __construct() {
+		$json_string       = get_option( self::OPTION_NAME );
+		$this->has_changed = false;
+
+		if ( $json_string ) {
+			$this->data_port_jobs = json_decode( $json_string, true );
+		} else {
+			$this->data_port_jobs = [];
+		}
+	}
+
+	/**
+	 * Initializes the data port manager.
+	 */
+	public function init() {
+		foreach ( $this->data_port_jobs as $job ) {
+			add_action( Sensei_Data_Port_Job::SCHEDULED_ACTION_NAME, [ $this, 'run_data_port_job' ] );
+		}
+
+		add_action( 'shutdown', [ $this, 'persist' ] );
+	}
+
+	/**
+	 * Runs a data port job.
+	 *
+	 * @param array $args  The arguments of the background job. Only the job_id is included.
+	 */
+	public function run_data_port_job( $args ) {
+
+		if ( empty( $args['job_id'] ) ) {
+			return;
+		}
+
+		$job = $this->get_job( $args['job_id'] );
+
+		if ( null !== $job ) {
+			Sensei_Scheduler::instance()->run( $job );
+		}
+	}
+
+	/**
+	 * Starts a data import job.
+	 *
+	 * @param int $user_id  The user which started the job.
+	 */
+	public function start_import_job( $user_id ) {
+		$job_id = md5( uniqid( '', true ) );
+
+		$this->has_changed      = true;
+		$this->data_port_jobs[] = [
+			'user_id' => $user_id,
+			'time'    => time(),
+			'handler' => 'Sensei_Import_Job',
+			'id'      => $job_id,
+		];
+
+		$job = new Sensei_Import_Job( $job_id );
+		Sensei_Scheduler::instance()->schedule_job( $job );
+	}
+
+	/**
+	 * Cancel a job.
+	 *
+	 * @param string $job_id  The job id.
+	 */
+	public function cancel_job( $job_id ) {
+		$job = $this->get_job( $job_id );
+
+		if ( null !== $job ) {
+			$job->clean_up();
+		}
+
+		$this->has_changed    = true;
+		$this->data_port_jobs = array_filter(
+			$this->data_port_jobs,
+			function ( $job ) use ( $job_id ) {
+				return $job['id'] !== $job_id;
+			}
+		);
+	}
+
+	/**
+	 * Cancel all pending jobs.
+	 */
+	public function cancel_all_jobs() {
+		foreach ( $this->data_port_jobs as $job ) {
+			if ( $job['handler'] instanceof \Sensei_Data_Port_Job ) {
```

Comment:
It doesn't look like `$job['handler']` is an object... am I missing it? If it is just a class name, I think we need to use `is_subclass_of( $job['handler'], Sensei_Data_Port_Job::class, true )`.

Reply:
I used `is_a` in `get_job` correctly but I somehow missed that one. 5728c4fdc5d6e077ac7ccd4caea34edaf5fea17f